File: rust/crates/chameleon-format-dialog/cpp/locale.cpp
#include "chameleon-format-dialog/cpp/includes/icu_includes.h"
#include "chameleon-format-dialog/cpp/locale.h"
#include "chameleon-format-dialog/cpp/includes/enums.h"
#include "chameleon-format-dialog/src/locale.cxx.h"

rust::Vec<LocaleInformation> get_locale_information(OutputFor output_for) {
    rust::Vec<LocaleInformation> locale_info;

    int32_t count;
    const Locale* list = NULL;
    UnicodeString result;

    switch (output_for) {
        case OutputFor::AllLocales: list = icu::Locale::getAvailableLocales(count); break;
        case OutputFor::NumberFormatLocales: list = icu::NumberFormat::getAvailableLocales(count); break;
        case OutputFor::DateFormatLocales: list = icu::DateFormat::getAvailableLocales(count); break;
        default: return locale_info;
    }

    LocaleInformation locale_info_item;
    for (int i = 0; i < count; i++) {
        locale_info_item.locale_name = list[i].getName();

        result = "";
        std::string tmp;
        list[i].getDisplayName(Locale::getRoot(), result);
        locale_info_item.locale_display_name = result.toUTF8String(tmp);

        locale_info.push_back(locale_info_item);
    }

    return locale_info;
}


const inline Locale get_default_locale() {
    return Locale::getRoot();
}

rust::String get_default_locale_name()
{
    return get_default_locale().getName();
}

rust::String get_default_locale_display_name()
{
    UnicodeString result = "";
    std::string tmp;
    get_default_locale().getDisplayName(Locale::getRoot(), result);
    return result.toUTF8String(tmp);
}

File: rust/crates/chameleon-format-dialog/src/locale.rs
#[cxx_qt::bridge]
mod ffi_icu {

    #[derive(Clone)]
    struct LocaleInformation {
        locale_name: String,
        locale_display_name: String,
    }

    unsafe extern "C++" {
        include!("chameleon-format-dialog/cpp/locale.h");

        type OutputFor = crate::format::OutputFor;
        fn get_locale_information(output_for: OutputFor) -> Vec<LocaleInformation>;
        fn get_default_locale_name() -> String;
        fn get_default_locale_display_name() -> String;

    }
}

pub use ffi_icu::{get_locale_information, LocaleInformation, get_default_locale_name, get_default_locale_display_name, OutputFor};


impl Default for LocaleInformation {
    fn default() -> Self {
        let locale_name = get_default_locale_name();
        let locale_display_name = get_default_locale_display_name();
        
        LocaleInformation {
            locale_name,
            locale_display_name,
        }
    }
}